### PR TITLE
[CONTP-394] use filters in tagger subscription

### PR DESCRIPTION
**Pull request by adel121**
### What does this PR do?

This PR refactors the subscription logic in the tagger so that it allows subscribers to subscribe with filters.

Subscription filters were implemented in a previous PR.

### Motivation

Allow subscribers to subscribe to tagger component using filters to constrain received notification to only what the subscriber is interested in.

### Describe how to test/QA your changes

Unit tests and E2E should be enough.

### CODE REVIEW

File: comp/core/tagger/taggerimpl/server/server.go [hunk 1]
```diff
@@ -27,9 +27,21 @@ const (
...
+// This unique ID is used to subscribe to the tagger
+// TODO: remove this struct when the protobuf of the stream request is updated to use filters.
+type streamIDManager int
+
+func (s *streamIDManager) getNextUniqueID() string {
+	id := fmt.Sprintf("stream-client-%d", *s)
+	*s++
+	return id
+}
+
```
Comment (adel121) [inline comp/core/tagger/taggerimpl/server/server.go:40]: This is just a temporary solution. 

Ideally, we will modify the protobuf message to include full data needed to perform the subscription to the tagger, including the subscription id. 

I preferred to keep modifying the protobuf to a separate PR to make it easier for review.

**Review by clamoriniere — COMMENTED** [review]
some initial comment about Component Interface modification.

Could you split the PR in 2:
* interface change
* implementation change

So like that we agree on the Interface change? before reviewing the implementation?

File: comp/core/tagger/component.go [hunk 2]
```diff
@@ -43,8 +43,8 @@ type Component interface {
 	Standard(entityID string) ([]string, error)
 	List() types.TaggerListResponse
 	GetEntity(entityID string) (*types.Entity, error)
-	Subscribe(cardinality types.TagCardinality) chan []types.EntityEvent
-	Unsubscribe(ch chan []types.EntityEvent)
+	Subscribe(subscriptionID string, filter *types.Filter) chan []types.EntityEvent
```
[the remark this answers is not in the document]
Comment (adel121) [reply comp/core/tagger/component.go:46]: Agree about clarifying purpose 👍 

but about format, I don't think there is any specific format. 
It is basically the same as in workloadmeta. 
Do you have any idea about special formats here?

File: comp/core/tagger/component.go [hunk 3]
```diff
@@ -43,8 +43,8 @@ type Component interface {
 	Standard(entityID string) ([]string, error)
 	List() types.TaggerListResponse
 	GetEntity(entityID string) (*types.Entity, error)
-	Subscribe(cardinality types.TagCardinality) chan []types.EntityEvent
-	Unsubscribe(ch chan []types.EntityEvent)
```
[the remark this answers is not in the document]
Comment (adel121) [reply comp/core/tagger/component.go:47]: A subscriber is now identified by a unique subscription id, and not by a channel. 

Ofcourse each subscriber still has a unique channel, the subscription data is organised by mapping a subscription id to its related information. 

We can still use the channel to identify the subscriber, but it will require an extra mapping step.

**Comment by adel121** [comment]
@clamoriniere Thank you for your insights. 

I updated the PR based on our discussion, and I split it into 2 commits.

File: comp/core/tagger/noopimpl/tagger.go [hunk 4]
```diff
@@ -73,12 +73,10 @@ func (n *noopTagger) GetEntity(string) (*types.Entity, error) {
 	return nil, nil
 }
 
-func (n *noopTagger) Subscribe(types.TagCardinality) chan []types.EntityEvent {
-	return make(chan []types.EntityEvent)
+func (n *noopTagger) Subscribe(string, *types.Filter) types.Subscription {
+	return nil
```
Comment (clamoriniere) [inline comp/core/tagger/noopimpl/tagger.go:77]: should you return a "noobSubscription" here or do you expect that the user is checking that the subscription is not nil?

IMO, since `Subscription` is an interface, `Subscribe()` should not return nil

File: comp/core/tagger/taggerimpl/server/server.go [hunk 5]
```diff
@@ -27,15 +28,30 @@ const (
...
+// streamIDManager is used to generate unique ID's for incoming streaming requests
+// This unique ID is used to subscribe to the tagger
+// TODO: remove this struct when the protobuf of the stream request is updated to use filters.
+type streamIDManager int32
+
+func (s *streamIDManager) getNextUniqueID() string {
+	id := fmt.Sprintf("stream-client-%d", *s)
+	atomic.AddInt32((*int32)(s), 1)
+	return id
+}
```
Comment (clamoriniere) [inline comp/core/tagger/taggerimpl/server/server.go:40]: IMO this solution is fine. 
I'm wondering which other alternative you have considered. have you think about using a UUID generator?

File: comp/core/tagger/taggerimpl/subscriber/subscription_manager.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	ch := make(chan []types.EntityEvent, bufferSize)
+	subscriber := &Subscriber{
+		filter:  *filter,
+		id:      id,
+		ch:      ch,
+		manager: sm,
+	}
+
+	sm.Lock()
+	sm.subscribers[id] = subscriber
```
Comment (clamoriniere) [inline comp/core/tagger/taggerimpl/subscriber/subscription_manager.go:58]: should add a check to know it the ID is already used? 
In that case should we return an error? and maybe change the function signature to `Subscribe(id string, filter *types.Filter, events []types.EntityEvent) (types.Subscription, error)` 

So like this you can also handle the case of the `noopTagger` implementation by returning an error that say "not implemented"

File: comp/core/tagger/taggerimpl/server/server.go [hunk 7]
```diff
@@ -27,15 +28,30 @@ const (
...
 )
 
+// streamIDManager is used to generate unique ID's for incoming streaming requests
+// This unique ID is used to subscribe to the tagger
+// TODO: remove this struct when the protobuf of the stream request is updated to use filters.
+type streamIDManager int32
+
+func (s *streamIDManager) getNextUniqueID() string {
+	id := fmt.Sprintf("stream-client-%d", *s)
+	atomic.AddInt32((*int32)(s), 1)
```
Comment (L3n41c) [inline comp/core/tagger/taggerimpl/server/server.go:38]: If the `streamIDManager` field is meant to be modified and accessed concurrently, it should use a safe `sync/atomic` type.

Using a “generic standard” type like `int32` leads to concurrency bugs like the one in the current implementation of `getNextUniqueID`:
* Whereas the increment done line 38 is thread-safe thanks to the use [`atomic.AddInt32`](https://pkg.go.dev/sync/atomic#AddInt32),
* The read done line 37 is **not** thread-safe because of the lack of use of [`atomic.LoadInt32`]( https://pkg.go.dev/sync/atomic#LoadInt32).

This first bug could be fixed by:
```suggestion
type streamIDManager int32

func (s *streamIDManager) getNextUniqueID() string {
	id := fmt.Sprintf("stream-client-%d", atomic.LoadInt32(s))
	atomic.AddInt32((*int32)(s), 1)
```

However, in order to avoid this kind of bug to be possible, it would be better to use a dedicated type that prevents non-atomic accesses:
```suggestion
type streamIDManager struct {
	atomic.Int64
}

func (s *streamIDManager) getNextUniqueID() string {
	id := fmt.Sprintf("stream-client-%d", s.Load())
	s.Add(1)
```

Although all accesses (both read and writes) to the `streamIDManager` field are now thread-safe, the `getNextUniqueID` function still suffer from a race condition.
It’s still possible to have two threads T1 and T2 calling `getNextUniqueID` and return exactly the same value with the following sequence:
```
  | value
  | of  s                      thread T1                                         thread T2
  |
  |  42
  |           id := fmt.Sprintf("stream-client-%d", s.Load())
  |  42
  |                                                             id := fmt.Sprintf("stream-client-%d", s.Load())
  |  42
  |           s.Add(1)
  |  43
  |                                                             s.Add(1)
  |  44
  |           return id
  |  44
  |                                                             return id
  |  44
  |
  V

time
```

With such an execution flow, with `s` being `42` before `getNextUniqueID` is being concurrently called by the two threads T1 and T2, `getNextUniqueID` would return exactly the same ID (`stream-client-42`) in both threads. Meaning that it wouldn’t be unique.
The following call would return `stream-client-44`.

It’s crucial that the read of the `s` value and its increment are atomic, which is not the case with any of the proposition above.

The most generic way to make the `fmt.Sprintf(…, s.Load())` and the `s.Add(1)` lines atomic is to protect them by a mutex.
But in this precise case, it would be overkill as we can atomically add and load the `s` value with:
```suggestion
type streamIDManager struct {
	atomic.Int64
}

func (s *streamIDManager) getNextUniqueID() string {
	id := fmt.Sprintf("stream-client-%d", s.Add(1))
```

File: comp/core/tagger/taggerimpl/subscriber/subscription_manager.go (hunk 6 above)
Comment (adel121) [reply comp/core/tagger/taggerimpl/subscriber/subscription_manager.go:58]: Agreed 👍 

Will update it

File: comp/core/tagger/taggerimpl/server/server.go (hunk 5 above)
Comment (adel121) [reply comp/core/tagger/taggerimpl/server/server.go:40]: UUID can work too.

But I didn't spend so much time thinking about this because it is only meant to be a temporary solution until we update the protobuf definition. 

Ideally, I think the cluster check runner (or generally, the subscriber), should provide the subscription ID.
It is in that place that I plan to use UUID or something similar. 

On the server side, we can then append some prefix to the id that is set by the subscribing component.

As an example, a cluster check runner can subscribe to the remote cluster tagger by specifying a filter and a uuid (say `77ashyyqqq`). The cluster tagger server would register the id as `cluster-check-runner-streamer-77ashyyqqq`.

File: comp/core/tagger/taggerimpl/server/server.go (hunk 7 above)
Comment (adel121) [reply comp/core/tagger/taggerimpl/server/server.go:38]: Awesome explanation !!! 🔥🔥🔥🔥🔥🔥

Thanks @L3n41c !!

**Review by clamoriniere — APPROVED** [review]
